### PR TITLE
Propagate DispatchError for benchmarks.

### PR DESCRIPTION
**Pull request by marcio-diaz**
Currently `dispatch_benchmark` is returning `None` if there is an error in the call.
This produces the unhelpful output of "No results." (cc @drahnr ).

In order to facilitate the benchmarking process this PR propagates `DispatchError` (e.g. producing output `Error: "TooManyFields"`).

### CODE REVIEW

File: primitives/runtime/src/lib.rs [hunk 1]
```diff
@@ -416,6 +416,13 @@ impl From<DispatchError> for &'static str {
 	}
 }
 
+impl From<DispatchError> for Vec<u8> {
+	fn from(err: DispatchError) -> Vec<u8> {
+		let s: &'static str = err.into();
+		s.as_bytes().to_vec()
+	}
```
Comment (marcio-diaz) [inline primitives/runtime/src/lib.rs]: Mmm, probably I should remove this and use it in `run_benchmarks`.

File: bin/node/runtime/src/lib.rs [hunk 2]
```diff
@@ -821,14 +821,14 @@ impl_runtime_apis! {
...
 			match module.as_slice() {
-				b"pallet-balances" | b"balances" => Balances::run_benchmark(extrinsic, steps, repeat).ok(),
-				b"pallet-identity" | b"identity" => Identity::run_benchmark(extrinsic, steps, repeat).ok(),
-				b"pallet-timestamp" | b"timestamp" => Timestamp::run_benchmark(extrinsic, steps, repeat).ok(),
-				_ => None,
+				b"pallet-balances" | b"balances" => Balances::run_benchmark(extrinsic, steps, repeat),
+				b"pallet-identity" | b"identity" => Identity::run_benchmark(extrinsic, steps, repeat),
+				b"pallet-timestamp" | b"timestamp" => Timestamp::run_benchmark(extrinsic, steps, repeat),
+				_ => Err("Benchmark not found for this pallet.".as_bytes().to_vec()),
 			}
```
Comment (bkchr) [inline bin/node/runtime/src/lib.rs]: Just put here a `.map_err(|e| e.encode())`

File: utils/frame/benchmarking-cli/src/lib.rs [hunk 3]
```diff
@@ -109,34 +109,37 @@ impl BenchmarkCmd {
...
-				let parameters = &result.0;
-				parameters.iter().for_each(|param| print!("{:?},", param.1));
-				// Print extrinsic time and storage root time
-				print!("{:?},{:?}\n", result.1, result.2);
-			});
-
-			eprintln!("Done.");
-		} else {
-			eprintln!("No Results.");
+		let results = <Result<Vec<BenchmarkResults>, Vec<u8>> as Decode>::decode(&mut &result[..])
```
Comment (bkchr) [inline utils/frame/benchmarking-cli/src/lib.rs]:
```suggestion
		let results = <Result<Vec<BenchmarkResults>, String> as Decode>::decode(&mut &result[..])
```

Comment (marcio-diaz) [reply utils/frame/benchmarking-cli/src/lib.rs]: I get `Error: "�Benchmark not found for this pallet."` if we don't decode twice. Other option is to encode the whole thing?, but maybe you don't want that, given your other comment of encoding just the error.

Comment (marcio-diaz) [reply utils/frame/benchmarking-cli/src/lib.rs]: I get `Error: "�Benchmark not found for this pallet."` if we don't decode twice. Other option is to encode the whole thing?, but maybe you don't want that, given your other comment of encoding just the error.

Comment (bkchr) [reply utils/frame/benchmarking-cli/src/lib.rs]: Ahh yeah! Sorry!

Could you please use [`RuntimeString`](https://github.com/paritytech/substrate/blob/879e28ae38643c028fa09c157e7b2714bcbf02a1/primitives/runtime/src/runtime_string.rs#L23)

(You also need to change the `Benchmark` trait to directly return `Result<BenchmarkResult, RuntimeString>`

Comment (marcio-diaz) [reply utils/frame/benchmarking-cli/src/lib.rs]: Nice, thanks!

File: utils/frame/benchmarking-cli/src/lib.rs [hunk 4]
```diff
@@ -109,34 +109,36 @@ impl BenchmarkCmd {
...
-				// Print extrinsic time and storage root time
-				print!("{:?},{:?}\n", result.1, result.2);
-			});
-
-			eprintln!("Done.");
-		} else {
-			eprintln!("No Results.");
+
+		let results = <Result<Vec<BenchmarkResults>, String> as Decode>::decode(&mut &result[..])
+			.expect("failed to decode benchmark results");
```
Comment (bkchr) [inline utils/frame/benchmarking-cli/src/lib.rs]:
```suggestion
			.map_err(|e| format!("Failed to decode benchmark results: {:?}", e))?;
```

I think that makes it cleaner if when it fails.

**Pull request merged**